torch/autograd/gradcheck.py

```diff
@@ -67,97 +137,31 @@ def get_numerical_jacobian(fn, input, target=None, eps=1e-3, grad_out=1.0):
-        elif ds_dx.is_complex():  # R -> C
-            # w_d = conj_w_d = 0.5 * ds_dx
-            # dL_dz_conj = 0.5 * [grad_out.conj() * ds_dx + grad_out * ds_dx.conj()]
-            #            = 0.5 * [grad_out.conj() * ds_dx + (grad_out.conj() * ds_dx).conj()]
-            #            = 0.5 * 2 * real(grad_out.conj() * ds_dx)
-            #            = real(grad_out.conj() * ds_dx)
-            d[d_idx] = torch.real(grad_out.conjugate() * ds_dx)
-        else:   # R -> R
-            d[d_idx] = ds_dx * grad_out
-
-    # TODO: compare structure
-    for x_tensor, d_tensor in zip(x_tensors, j_tensors):
-        if x_tensor.is_sparse:
-            def get_stride(size):
-                dim = len(size)
-                tmp = 1
-                stride = [0] * dim
-                for i in reversed(range(dim)):
-                    stride[i] = tmp
-                    tmp *= size[i]
-                return stride
-
-            x_nnz = x_tensor._nnz()
-            x_size = list(x_tensor.size())
-            x_indices = x_tensor._indices().t()
-            x_values = x_tensor._values()
-            x_stride = get_stride(x_size)
-
-            # Use .data here to get around the version check
-            x_values = x_values.data
-
-            for i in range(x_nnz):
-                x_value = x_values[i]
-                for x_idx in product(*[range(m) for m in x_values.size()[1:]]):
-                    indices = x_indices[i].tolist() + list(x_idx)
-                    d_idx = sum(indices[k] * x_stride[k] for k in range(len(x_size)))
-                    update_jacobians(x_value, x_idx, d_tensor, d_idx)
-        elif x_tensor.layout == torch._mkldnn:  # type: ignore
-            # Use .data here to get around the version check
-            x_tensor = x_tensor.data
-            if len(input) != 1:
-                raise ValueError('gradcheck currently only supports functions with 1 input, but got: ',
-                                 len(input))
-            for d_idx, x_idx in enumerate(product(*[range(m) for m in x_tensor.size()])):
-                # this is really inefficient, but without indexing implemented, there's
-                # not really a better way than converting back and forth
-                x_tensor_dense = x_tensor.to_dense()
-                update_jacobians(x_tensor_dense, x_idx, d_tensor, d_idx, is_mkldnn=True)
-        else:
-            # Use .data here to get around the version check
-            x_tensor = x_tensor.data
-            for d_idx, x_idx in enumerate(product(*[range(m) for m in x_tensor.size()])):
-                update_jacobians(x_tensor, x_idx, d_tensor, d_idx)
-
+    for x_tensor, d in zip(x_tensors, j_tensors):
+        is_mkldnn = x_tensor.layout == torch._mkldnn  # type: ignore # no attr _mkldnn
+        for x, idx, d_idx in iter_tensor(x_tensor):
+            # compute_jacobian only works for pure real or pure imaginary delta
```

Comment:
What "compute_jacobian" is this referring to?

Reply:
Hmm, that comment should probably be updated to "computing the jacobian". It should refer to computing the jacobian in general, though the reason why I don't quite understand yet. It might be somewhat related to the fact that we are trying to compute the derivatives of non-analytic functions.